engine/src/main/java/org/pentaho/di/repository/kdr/delegates/KettleDatabaseRepositoryMetaStoreDelegate.java

```diff
@@ -0,0 +1,516 @@
+        throw new MetaStoreException( "Unable to find namespace with name '" + namespace + "'" );
+      }
+      return namespaceId;
+    } catch ( KettleException e ) {
+      throw new MetaStoreException( "Unable to get id of namespace '" + namespace + "'", e );
+    }
+  }
+
+  public synchronized LongObjectId getElementTypeId( LongObjectId namespaceId, String elementTypeName ) throws KettleException {
+    return repository.connectionDelegate.getIDWithValue(
+      quoteTable( KettleDatabaseRepository.TABLE_R_ELEMENT_TYPE ),
+      quote( KettleDatabaseRepository.FIELD_ELEMENT_TYPE_ID_ELEMENT_TYPE ),
+      quote( KettleDatabaseRepository.FIELD_ELEMENT_TYPE_NAME ), elementTypeName,
+      new String[] { quote( KettleDatabaseRepository.FIELD_ELEMENT_TYPE_ID_NAMESPACE ), },
+      new ObjectId[] { namespaceId, } );
+  }
+
+  public synchronized LongObjectId getElementId( LongObjectId elementTypeId, String elementName ) throws KettleException {
+    return repository.connectionDelegate.getIDWithValue(
+      quoteTable( KettleDatabaseRepository.TABLE_R_ELEMENT ),
+      quote( KettleDatabaseRepository.FIELD_ELEMENT_ID_ELEMENT ),
+      quote( KettleDatabaseRepository.FIELD_ELEMENT_NAME ), elementName,
+      new String[] { quote( KettleDatabaseRepository.FIELD_ELEMENT_ID_ELEMENT_TYPE ), },
+      new ObjectId[] { elementTypeId, } );
+  }
+
+  public RowMetaAndData getElementType( LongObjectId elementTypeId ) throws KettleException {
+    return repository.connectionDelegate.getOneRow(
+      quoteTable( KettleDatabaseRepository.TABLE_R_ELEMENT_TYPE ),
+      quote( KettleDatabaseRepository.FIELD_ELEMENT_TYPE_ID_ELEMENT_TYPE ), elementTypeId );
+  }
+
+  public RowMetaAndData getElement( LongObjectId elementId ) throws KettleException {
+    return repository.connectionDelegate.getOneRow(
+      quoteTable( KettleDatabaseRepository.TABLE_R_ELEMENT ),
+      quote( KettleDatabaseRepository.FIELD_ELEMENT_ID_ELEMENT ), elementId );
+  }
+
+  public RowMetaAndData getElementAttribute( LongObjectId elementAttributeId ) throws KettleException {
+    return repository.connectionDelegate.getOneRow(
+      quoteTable( KettleDatabaseRepository.TABLE_R_ELEMENT_ATTRIBUTE ),
+      quote( KettleDatabaseRepository.FIELD_ELEMENT_ATTRIBUTE_ID_ELEMENT_ATTRIBUTE ), elementAttributeId );
+  }
+
+  public Collection<RowMetaAndData> getNamespaces() throws KettleDatabaseException, KettleValueException {
+    List<RowMetaAndData> attrs = new ArrayList<RowMetaAndData>();
+
+    String sql = "SELECT * FROM " + quoteTable( KettleDatabaseRepository.TABLE_R_NAMESPACE );
+
+    List<Object[]> rows = repository.connectionDelegate.getRows( sql, 0 );
+    for ( Object[] row : rows ) {
+      RowMetaAndData rowWithMeta = new RowMetaAndData( repository.connectionDelegate.getReturnRowMeta(), row );
+      long id = rowWithMeta.getInteger( quote( KettleDatabaseRepository.FIELD_NAMESPACE_ID_NAMESPACE ), 0 );
+      if ( id > 0 ) {
+        attrs.add( rowWithMeta );
+      }
+    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 126:132, 146:152, 166:172, 191:199.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.